kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/AsyncProcessor.java

```diff
@@ -0,0 +1,456 @@
+  }
+
+  @Override
+  public void process(final FixedKeyRecord<KIn, VIn> record) {
+    final AsyncEvent newEvent = new AsyncEvent(
+        logPrefix,
+        record,
+        taskId.partition(),
+        originalContext.recordContext(),
+        originalContext.currentStreamTimeMs(),
+        originalContext.currentSystemTimeMs(),
+        () -> userFixedKeyProcessor.process(record)
+    );
+
+    processNewAsyncEvent(newEvent);
+  }
+
+  private void processNewAsyncEvent(final AsyncEvent event) {
+    pendingEvents.add(event);
+    schedulingQueue.offer(event);
+
+    executeAvailableEvents();
+  }
+
+  @Override
+  public void close() {
+    if (!pendingEvents.isEmpty()) {
+      // This doesn't necessarily indicate an issue, it just should only ever
+      // happen if the task is closed dirty, but unfortunately we can't tell
+      // from here whether that was the case. Log a warning here so that it's
+      // possible to determine whether something went wrong or not by looking
+      // at the complete logs for the task/thread
+      log.warn("Closing async processor with {} in-flight events, this should only "
+                   + "happen if the task was shut down dirty and not flushed/committed "
+                   + "prior to being closed", pendingEvents.size());
+    }
+
+    threadPool.removeProcessor(taskId.partition());
+    unregisterFlushListenerForStoreBuilders(
+        streamThreadName,
+        taskId.partition(),
+        connectedStoreBuilders.values()
+    );
+
+    // Tell the user context to turn off processing mode so it knows to expect no
+    // further calls from an AsyncThread after this
+    userContext.endProcessingMode();
+    userProcessor.close();
+  }
+
+  private static void unregisterFlushListenerForStoreBuilders(
+      final String streamThreadName,
+      final int partition,
+      final Collection<AsyncStoreBuilder<?>> asyncStoreBuilders
+  ) {
+    for (final AsyncStoreBuilder<?> builder : asyncStoreBuilders) {
+      builder.unregisterFlushListenerForPartition(streamThreadName, partition);
+    }
+  }
+
+  private static void registerFlushListenerForStoreBuilders(
+      final String streamThreadName,
+      final int partition,
+      final Collection<AsyncStoreBuilder<?>> asyncStoreBuilders,
+      final AsyncFlushListener flushAllAsyncEvents
+  ) {
+    for (final AsyncStoreBuilder<?> builder : asyncStoreBuilders) {
+      builder.registerFlushListenerForPartition(streamThreadName, partition, flushAllAsyncEvents);
+    }
+  }
+
+  /**
+   * Block on all pending records to be scheduled, executed, and fully complete processing through
+   * the topology, as well as all state store operations to be applied. Called at the beginning of
+   * each commit, similar to #flushCache
+   */
+  public void flushAndAwaitCompletion() {
+    while (!isCompleted()) {
+
+      drainSchedulingQueue();
+      drainFinalizingQueue();
+
+      // TODO: use a Condition to avoid busy waiting
```

Comment:
this is definitely worth doing soon

Reply:
agreed (finished this up a little while ago)